rosbag2_cpp/src/rosbag2_cpp/reindexer.cpp

```diff
@@ -0,0 +1,252 @@
+// Copyright 2020 DCS Corporation, All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// DISTRIBUTION A. Approved for public release; distribution unlimited.
+// OPSEC #4584.
+//
+// Delivered to the U.S. Government with Unlimited Rights, as defined in DFARS
+// Part 252.227-7013 or 7014 (Feb 2014).
+//
+// This notice must appear in all copies of this file and its derivatives.
+
+#include <algorithm>
+#include <iostream>
+#include <map>
+#include <memory>
+#include <regex>
+#include <stdexcept>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "rcpputils/asserts.hpp"
+#include "rcpputils/filesystem_helper.hpp"
+
+#include "rcutils/filesystem.h"
+
+#include "rosbag2_cpp/logging.hpp"
+#include "rosbag2_cpp/reader.hpp"
+#include "rosbag2_cpp/reindexer.hpp"
+
+#include "rosbag2_storage/storage_options.hpp"
+
+#ifdef WIN32
+// Import windows filesystem functionality
+#include <windows.h>
+#include <tchar.h>
+#include <stdio.h>
+#else
+// We're on a UNIX system. Import their filesystem stuff instead
+#include <dirent.h>
+#endif
+
+
+namespace rosbag2_cpp
+{
+Reindexer::Reindexer(
+  std::unique_ptr<rosbag2_storage::StorageFactoryInterface> storage_factory,
+  std::unique_ptr<rosbag2_storage::MetadataIo> metadata_io)
+: storage_factory_(std::move(storage_factory)),
+  metadata_io_(std::move(metadata_io)) {}
+
+Reindexer::~Reindexer() {}
+
+bool Reindexer::compare_relative_file(
+  const rcpputils::fs::path & first_path, const rcpputils::fs::path & second_path)
+{
+  std::regex regex_rule(".*_(\\d+)\\.db3", std::regex_constants::ECMAScript);
```

Comment:
This is hardcoded to db3 files? Let's have the storage implementation give us the expected extension. 

Or for a looser requirement, we could generalize the regex to 
```
.*_(\d+)\.([a-z0-9+])
```

So that we capture that it's "name_{N}.{EXT}"

Reply:
How do I get the storage implementation give the extension? I just want to make sure that we don't pick up something like a `.db3-wal` file, and mis-collect it as a relative file...

Reply:
I think a new pure-virtual API in https://github.com/ros2/rosbag2/blob/master/rosbag2_storage/include/rosbag2_storage/storage_interfaces/base_io_interface.hpp would make sense - `virtual std::string get_default_bagfile_extension() = 0;`

Reply:
Unfortunately, I'm running into a problem getting this to work. I can create the API just fine, but actually using it in the Reindexer is another story.

The Reindexer uses a Reader instance in order to access the files, rather than using the storage directly. This creates a couple problems: The storage within the Reader is `protected` which means that I can't get to it from my Reader instance. Trying to get around this problem by creating a function that queries the storage for its extension makes me run into another "interesting" issue: until the reader opens a file, `storage_` doesn't exist!

Which means, in order to determine which files I'm allowed to open, I have to open a file...

The only way I can think of to get around this problem would be to create a function that queries the `rosbag2_storage::StorageOptions` that would be used to open the file, and query its storage_id. Then I construct the extension from that. But that opens a slight risk of code duplication, and I'm not sure how I'd be able to make it extendable to other bag types.

Any ideas?